File: CS/GridColumnHeight.Module.Win/GridHeaderHeightHelper.cs
using System;
using System.Drawing;
using System.Text;
using DevExpress.ExpressApp;
using DevExpress.Utils.Drawing;
using DevExpress.XtraEditors.Drawing;
using DevExpress.XtraGrid.Drawing;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.XtraGrid.Views.Grid.ViewInfo;
using DevExpress.XtraGrid;
using DevExpress.ExpressApp.Win.Editors;

namespace XAFCommunity.Win
{
    public static class GridHeaderHeightHelper
    {
        internal static Schema GetGridHeaderHeightSchema()
        {
            const string schema =
                @"<Element Name=""Application"">
	                <Element Name=""Views"">
		                <Element Name=""ListView"">
                            <Attribute Name=""HeaderRows"" IsNewNode=""True""/>
		                </Element>
		                <Element Name=""LookupListView"">
                            <Attribute Name=""HeaderRows"" IsNewNode=""True""/>
		                </Element>
	                </Element>
                </Element>";
            
            return new Schema(new DictionaryXmlReader().ReadFromString(schema));
        }

        /// <summary>
        /// Computes a suitable height for the column header based upon the number of column header lines required.
        /// </summary>
        /// <param name="view"></param>
        /// <param name="minHeaderHeight"></param>
        /// <param name="lineCount"></param>
        /// <returns></returns>
        internal static int CalcMinColumnRowHeight(GridView view, int minHeaderHeight, int lineCount)
        {
            int maxY = 0;
            minHeaderHeight = Math.Max(minHeaderHeight, 12);

            if (view != null && lineCount > 1)
            {
                GridViewInfo viewInfo = view.GetViewInfo() as GridViewInfo;
                Graphics graphic = viewInfo.GInfo.AddGraphics(null);

                try
                {
                    GridColumnInfoArgs e = new GridColumnInfoArgs(viewInfo.GInfo.Cache, null);
                    e.InnerElements.Add(new DrawElementInfo(new GlyphElementPainter(), new GlyphElementInfoArgs(view.Images, 0, null), StringAlignment.Near));
                    if (view.CanShowFilterButton(null))
                        e.InnerElements.Add(viewInfo.Painter.ElementsPainter.FilterButton, new GridFilterButtonInfoArgs());
                    e.SetAppearance(viewInfo.PaintAppearance.HeaderPanel);
                    maxY = viewInfo.Painter.ElementsPainter.Column.CalcObjectMinBounds(e).Height;
                    if (view.OptionsView.AllowHtmlDrawHeaders)
                    {
                        string dummyCaption = new StringBuilder("Ûj").Insert(0, "Ûj\n", lineCount - 1).ToString();
                        e.UseHtmlTextDraw = true;
                        e.Caption = dummyCaption;
                        maxY = Math.Max(maxY, viewInfo.Painter.ElementsPainter.Column.CalcObjectMinBounds(e).Height);
                    }
                }
                finally
                {
                    viewInfo.GInfo.ReleaseGraphics();
                }
            }
            return Math.Max(maxY, minHeaderHeight);
        }

        internal static void SetGridEditorHeaderHeight(GridListEditor listEditor, int lineCount)
        {
            if (listEditor != null)
            {
                GridControl gridControl = listEditor.Control as GridControl;
                if (gridControl != null)
                {
                    GridView gridView = gridControl.FocusedView as GridView;
                    if (gridView != null)
                    {
                        gridView.OptionsView.AllowHtmlDrawHeaders = true;
                        gridView.Appearance.HeaderPanel.TextOptions.WordWrap = DevExpress.Utils.WordWrap.Wrap;
                        int columnHeaderHeight =
                            GridHeaderHeightHelper.CalcMinColumnRowHeight(gridView, -1, lineCount);
                        gridView.ColumnPanelRowHeight = columnHeaderHeight;
                    }
                }
            }
        }

    }
}

File: VB/GridColumnHeight.Module.Win/GridHeaderHeightHelper.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Drawing
Imports System.Text
Imports DevExpress.ExpressApp
Imports DevExpress.Utils.Drawing
Imports DevExpress.XtraEditors.Drawing
Imports DevExpress.XtraGrid.Drawing
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.XtraGrid.Views.Grid.ViewInfo
Imports DevExpress.XtraGrid
Imports DevExpress.ExpressApp.Win.Editors

Namespace XAFCommunity.Win
	Public NotInheritable Class GridHeaderHeightHelper
		Private Sub New()
		End Sub
		Friend Shared Function GetGridHeaderHeightSchema() As Schema
			Const schema As String = "<Element Name=""Application"">" & ControlChars.CrLf & "	                <Element Name=""Views"">" & ControlChars.CrLf & "		                <Element Name=""ListView"">" & ControlChars.CrLf & "                            <Attribute Name=""HeaderRows"" IsNewNode=""True""/>" & ControlChars.CrLf & "		                </Element>" & ControlChars.CrLf & "		                <Element Name=""LookupListView"">" & ControlChars.CrLf & "                            <Attribute Name=""HeaderRows"" IsNewNode=""True""/>" & ControlChars.CrLf & "		                </Element>" & ControlChars.CrLf & "	                </Element>" & ControlChars.CrLf & "                </Element>"

			Return New Schema(New DictionaryXmlReader().ReadFromString(schema))
		End Function

		''' <summary>
		''' Computes a suitable height for the column header based upon the number of column header lines required.
		''' </summary>
		''' <param name="view"></param>
		''' <param name="minHeaderHeight"></param>
		''' <param name="lineCount"></param>
		''' <returns></returns>
		Friend Shared Function CalcMinColumnRowHeight(ByVal view As GridView, ByVal minHeaderHeight As Integer, ByVal lineCount As Integer) As Integer
			Dim maxY As Integer = 0
			minHeaderHeight = Math.Max(minHeaderHeight, 12)

			If view IsNot Nothing AndAlso lineCount > 1 Then
				Dim viewInfo As GridViewInfo = TryCast(view.GetViewInfo(), GridViewInfo)
				Dim graphic As Graphics = viewInfo.GInfo.AddGraphics(Nothing)

				Try
					Dim e As New GridColumnInfoArgs(viewInfo.GInfo.Cache, Nothing)
					e.InnerElements.Add(New DrawElementInfo(New GlyphElementPainter(), New GlyphElementInfoArgs(view.Images, 0, Nothing), StringAlignment.Near))
					If view.CanShowFilterButton(Nothing) Then
						e.InnerElements.Add(viewInfo.Painter.ElementsPainter.FilterButton, New GridFilterButtonInfoArgs())
					End If
					e.SetAppearance(viewInfo.PaintAppearance.HeaderPanel)
					maxY = viewInfo.Painter.ElementsPainter.Column.CalcObjectMinBounds(e).Height
					If view.OptionsView.AllowHtmlDrawHeaders Then
						Dim dummyCaption As String = New StringBuilder("Ûj").Insert(0, "Ûj" & Constants.vbLf, lineCount - 1).ToString()
						e.UseHtmlTextDraw = True
						e.Caption = dummyCaption
						maxY = Math.Max(maxY, viewInfo.Painter.ElementsPainter.Column.CalcObjectMinBounds(e).Height)
					End If
				Finally
					viewInfo.GInfo.ReleaseGraphics()
				End Try
			End If
			Return Math.Max(maxY, minHeaderHeight)
		End Function

		Friend Shared Sub SetGridEditorHeaderHeight(ByVal listEditor As GridListEditor, ByVal lineCount As Integer)
			If listEditor IsNot Nothing Then
				Dim gridControl As GridControl = TryCast(listEditor.Control, GridControl)
				If gridControl IsNot Nothing Then
					Dim gridView As GridView = TryCast(gridControl.FocusedView, GridView)
					If gridView IsNot Nothing Then
						gridView.OptionsView.AllowHtmlDrawHeaders = True
						gridView.Appearance.HeaderPanel.TextOptions.WordWrap = DevExpress.Utils.WordWrap.Wrap
						Dim columnHeaderHeight As Integer = GridHeaderHeightHelper.CalcMinColumnRowHeight(gridView, -1, lineCount)
						gridView.ColumnPanelRowHeight = columnHeaderHeight
					End If
				End If
			End If
		End Sub

	End Class
End Namespace
